### DISCUSSION TITLE
cpu idle not balance in 3 node small cluster

### DISCUSSION THREAD

**Discussion by Estheriii** [emqx/emqx#11616] [General]
### What happened?

1. Start emqx on three Ubuntu 20.04 machines to form a 3-node emqx cluster, with all node roles being core roles。
2. 1.35 million connections to the cluster, evenly distributed across each node
3. The CPU usage rate of each virtual machine is uneven, with significant differences

### What did you expect to happen?

The CPU usage rate of each virtual machine should be similar

### How can we reproduce it (as minimally and precisely as possible)?

1. Start emqx on three Ubuntu 20.04 virtual machines to form a 3-node emqx cluster, with all node roles being core roles。
2. 1.35 million connections to the cluster

### Anything else we need to know?

node1：
<img width="333" alt="image" src="https://github.com/emqx/emqx/assets/32813533/c454d300-bd4b-441a-a279-61d9ac7c9940">
node2：
<img width="324" alt="image" src="https://github.com/emqx/emqx/assets/32813533/dbf45d71-84d4-49d1-89cc-28e7bc4b3de3">
node3:
<img width="335" alt="image" src="https://github.com/emqx/emqx/assets/32813533/6acdd9f8-59f3-45a0-a939-c3238da63368">

conf file:
[conf.zip](https://github.com/emqx/emqx/files/12534688/conf.zip)

### EMQX version

<details>

```console
$ ./bin/emqx_ctl broker
# paste output here
sysdescr  : EMQX
version   : 5.0.21-gb2fc56ad
datetime  : 2023-09-06T07:28:05.609882959+00:00
uptime    : 7 days, 6 hours, 26 minutes, 11 seconds

```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
NAME="Ubuntu"
VERSION="20.04 LTS (Focal Fossa)"

$ uname -a
# paste output here
Linux jinpeng11 5.4.0-26-generic #30-Ubuntu SMP Mon Apr 20 16:58:30 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Log files

<details>

</details>

**Comment by savonarola** [comment]
Hello!
It's quite difficult to identify the reason of such behavior having this little information.
However, it can be seen that on the node with the highest CPU consumption (node1) system times are significantly greater. This may occur e.g. because of excess IO load on the node.
So could you:
* Also compare IO resource consumption on nodes (both FS IO and network IO).
* Investigate/provide logs of EMQX on the nodes.

**Comment by Estheriii** [comment]
> Hello! It's quite difficult to identify the reason of such behavior having this little information. However, it can be seen that on the node with the highest CPU consumption (node1) system times are significantly greater. This may occur e.g. because of excess IO load on the node. So could you:
> 
> * Also compare IO resource consumption on nodes (both FS IO and network IO).
> * Investigate/provide logs of EMQX on the nodes.

how can i get FS IO and network IO from emqx, what is the command?
This phenomenon is so common that I think you can form your own cluster to reproduce it

**Comment by id** [comment]
@Estheriii FS IO and network IO can be obtained with, for example, [iotop](https://www.man7.org/linux/man-pages/man8/iotop.8.html) and [iftop](https://linux.die.net/man/8/iftop).

Can you confirm that connections are indeed evenly distributed between the nodes?

**Comment by qzhuyan** [comment]
do you have full control of the client? 
do all the clients emit the same traffic to EMQX?

**Comment by ieQu1** [comment]
We need to know more about the client behavior. 
Just to give an example: if you have N clients that only publish messages and one client that subscribes to `#`, then the node that handles the subscriber will have to do more work, even though the number of client is more or less equal on all nodes.

**Comment by Estheriii** [comment]
> We need to know more about the client behavior. Just to give an example: if you have N clients that only publish messages and one client that subscribes to `#`, then the node that handles the subscriber will have to do more work, even though the number of client is more or less equal on all nodes.

client connect to the node and subscribe a topic : /device/{deviceId}/down 
which deviceId is random string
and publish message to /device/{deviceId}/up    total outgoing rate 100tps
in this issue no client subscribe to /device/{deviceId}/up
and no message through topic   /device/{deviceId}/down

**Reply by qzhuyan** [reply]
~~what about the subscribers?~~

are all the clients connected from different networks (different region, operators)>

could you check the Prometheus metrics? see if there are any differences among nodes.

**Reply by Estheriii** [reply]
clients from different ips, but some of them are the same
i have checked prometheus metrics nothing different
